compiler/src/iree/compiler/Codegen/Common/TileSwizzle.h

```diff
@@ -9,27 +9,74 @@
 
 #include <cstdint>
 #include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
 
 namespace mlir::iree_compiler {
 
 // Metadata for a swizzle, that is, an (expand_shape -> transposition)
 // pair of ops performing a change of layout within the tiles. This is used
 // on GPU, where the tiles themselves can have an arbitrary layout.
 struct TileSwizzle {
+  struct Dim {
+    // Describes what varies across this dimension.
+    enum class Kind : int8_t {
+      // This dimension is internal to one intrinsic on one thread. This
+      // is only seen for intrinsic operands that are themselves vectors.
+      // For example, with AMD MFMA, for the MFMA_F32_16x16x4_F32 intrinsic,
+      // the C-matrix operand is a vector of 4 floats already at the level of
+      // one intrinsic on one thread. That dimension of size 4 is 'Internal'.
+      Internal,
+      // This dimension is internal to one intrinsic, but is across threads.
+      // For example, with AMD MFMA, for the MFMA_F32_16x16x4_F32 intrinsic,
+      // the A-matrix tile has shape 16x4, and these two dimensions of size 16
+      // and 4 are 'CrossThread': neither is visible at the single-thread level
+      // (in the intrinsic itself, the A-matrix operand is a single scalar) but
+      // as we move along these dimensions, we are moving over the 64 threads
+      // of the subgroup.
+      //
+      // Another example of cross-thread dimensions is in kernels that are
+      // "unrolled" across subgroups. Such dimensions are cross-subgroup, so in
+      // particular they are cross-thread.
+      CrossThread,
+      // This dimensions is across intrinsics. By definition, this only occurs
+      // in kernels that have unrolling dimensions, so each thread executes
+      // multiple intrinsics. The 'CrossIntrinsic' dimensions then index over
+      // the intrinsics to be executed by each thread.
+      CrossIntrinsic
+    };
+
+    Kind kind = Kind::Internal;
+
+    // The size of the dimension.
+    int16_t size = 0;
+  };
+  static_assert(sizeof(Dim) == 4);
```

Comment:
Why do we need the assertion?

Reply:
i was just careful to minimize sizeof. can drop the static_assert now.